src/paginationCalls/pageDescriptor.ts

```diff
@@ -103,6 +116,108 @@ export class PageDescriptor implements Descriptor {
+    const responsePageTokenFieldName = self.responsePageTokenField;
+    const resourceField = self.resourceField;
+    const getNextPageRequest = self.getNextPageRequest;
+    const asyncIterable = {
+      [Symbol.asyncIterator]() {
+        const funcPromise = new Promise((resolve, reject) => {
+          self.resolveFunction = resolve;
+        });
+        const requestPromise = new Promise((resolve, reject) => {
+          self.resolveRequest = resolve;
+        });
+        const cache: Array<{}> = [];
+        let nextPageRequest: RequestType | null = {};
+        let firstCall = true;
+        return {
+          async next() {
+            const ongoingCall = new call.OngoingCallPromise(options.promise);
+            const func = (await funcPromise) as SimpleCallbackFunction;
+            const request = (await requestPromise) as RequestType;
+            if (cache.length > 0) {
+              const value = cache.shift();
+              return Promise.resolve({done: false, value});
+            }
+            if (!firstCall && !nextPageRequest) {
+              return Promise.resolve({done: true, value: -1});
+            }
+            nextPageRequest = await getNextPageRequest(
+              cache,
+              func,
+              firstCall ? request : nextPageRequest!,
+              ongoingCall
+            );
+            firstCall = false;
+            if (cache.length === 0) {
+              nextPageRequest = null;
+              return Promise.resolve({done: true, value: -1});
+            }
+            return Promise.resolve({done: false, value: cache.shift()});
+          },
+        };
+      },
+    };
+    return asyncIterable; // return iterable
+  }
+
+  async getNextPageRequest(
+    cache: Array<{}>,
+    func: SimpleCallbackFunction,
+    request: RequestType,
+    ongoingCall: call.OngoingCallPromise
+  ): Promise<RequestType | null> {
+    const requestPageTokenFieldName = this.requestPageTokenField;
+    const responsePageTokenFieldName = this.responsePageTokenField;
+    const resourceField = this.resourceField;
+    ongoingCall.call(func, request);
+    let nextPageRequest = null;
+    const [response, nextRequest, rawResponse] = await ongoingCall.promise;
+    const pageToken = (response as ResponseType)[responsePageTokenFieldName];
+    if (pageToken) {
+      nextPageRequest = Object.assign({}, request);
+      nextPageRequest[requestPageTokenFieldName] = pageToken;
+    }
+    const responses = (response as ResponseType)[resourceField];
+    cache.push(...responses);
+    return nextPageRequest;
+  }
+
+  resolveParams(request: RequestType, func: GaxCall, settings: CallSettings) {
+    if (settings.pageToken) {
+      request[this.requestPageTokenField] = settings.pageToken;
+    }
+    if (settings.pageSize) {
+      request[this.requestPageSizeField!] = settings.pageSize;
+    }
+    this.resolveRequest(request);
+    this.resolveFunction(func);
```

Comment:
So... do we need `resolveFunction` at all? Seems like `resolveRequest` is totally enough.

Reply:
I think we need... since when creating iterable, we do not have function and request available, so we might have to wait for both of them to be resolved and then making real request?

Reply:
will try resolve both of them at the same time, thanks for the suggestion